### DISCUSSION TITLE
Can we rollback our database transactions in strapi?

### DISCUSSION THREAD

**Discussion by sanojsilva** [strapi/strapi#6437] [Legacy Topics]
Hey is there a way to rollback database transactions if something goes wrong?

**Comment by lauriejim** [comment]
Hello, this is not possible to do that.
Can you please give me more detail about the way you would like that works in Strapi.
This will help us. Thank you.

**Reply by sonlamnguyen** [reply]
I have the same problem. I want to do serial jobs with DB.

Create a post.
Create a comment follow to post and occur errors.
I want to rollback post.
So, I don't find a document about transactions. Please, can you help me?

**Comment by MattieBelt** [comment]
@sonlamnguyen Maybe you could use the bookshelf instance if using a SQL DB. I have no experience in using [Bookshelf transactions](https://bookshelfjs.org/api.html#Bookshelf-instance-transaction).

`strapi.connections.default.transaction();`

**Comment by sanojsilva** [comment]
I have used the following way to manage db transactions.
```javascript
 const res = await strapi.connections.default.transaction(async (trx) => {
    await trx("voters").insert(result.data);
    return true;
 });
```

check out this link for more information 
[http://knexjs.org/#Transactions](http://knexjs.org/#Transactions)

**Reply by sonlamnguyen** [reply]
@sanojsilva Thank you very much!. you saved my life

**Comment by luwes** [comment]
the pattern used in strapi core seems to be something like the below.
see here, https://github.com/strapi/strapi/blob/97922f99ae858e1e3f4f792bb8d6a829adb82649/packages/strapi-connector-bookshelf/lib/queries.js#L39-L39

an example of my project:

```js
      await strapi.connections.default.transaction(async (transacting) => {

        mediaFile = await strapi.services['media-file'].getOrCreate(
          { md5 },
          mediaFileData,
          { transacting }
        );

        const clip = await strapi.services.clip.findOne({ id }, null, {
          transacting,
        });

        await strapi.services.clip.update(
          { id },
          {
            mediaFiles: _.uniq([...clip.mediaFiles, mediaFile.id]),
          },
          { transacting }
        );
      });
```


`getOrCreate` is one of my custom service functions by the way, it's not added by default.

```js
  async getOrCreate(params, data, { transacting } = {}) {
    let entry = await this.findOne(params, null, { transacting });
    if (!entry) {
      entry = await this.create(data, { transacting });
    }
    return entry;
  }
```

**Reply by ps-rschantl** [reply]
@luwes Did you test your implementation, because it looks like Strapi doesn't pass the transaction to the bookshelf-connector functions when you use the entity services?

So e.g. this
```javascript
strapi.services[MODEL_NAME].create(ctx.request.body, { transacting })
```
doesn't rollback but if I directly go to the bookshelf-connector like this
```javascript
strapi.query(MODEL_NAME).create(ctx.request.body, { transacting })
```
it works.

**Reply by luwes** [reply]
good catch, I don't think I tested 🤦 

there's also strapi.entityService.create, not sure what the difference is there
https://github.com/strapi/strapi/blob/9b62d98b459f0482c16e4b66cfd39638dac6c565/packages/strapi/lib/core-api/service.js#L74-L74

**Reply by emahuni** [reply]
@luwes could you please shed light on how you extended every model's services to include your custom getOrCreate service

**Reply by emahuni** [reply]
this transaction method doesn't work in current version of Strapi version 3.3.x

**Comment by derrickmehaffy** [comment]
Locked as moved to: https://forum.strapi.io/t/can-we-rollback-our-database-transactions-in-strapi-6437/1497
